src/trainer.py

```diff
@@ -0,0 +1,222 @@
+import itertools
+import torch
+import torch.nn as nn
+import models
+from utils.utils import mask_generator, weights_init
+from utils.utils import LR_lambda
+
+# TODO clean up & add comments & rename some variables
+# TODO maybe create dedicated discriminator's optimizers
+# TODO what should disc_f2s and disc_s2f return
+# TODO queue mask and buffer classes
+
+
+class Trainer:
+    def __init__(self, opt) -> None:
+        self.opt = opt
+        # networks
+        self.generator_shadow_to_free = models.Deshadower(
+            in_channels=opt.in_channels, out_channels=opt.out_channels
+        )
+        self.generator_free_to_shadow = models.Shadower(
+            in_channels=opt.in_channels, out_channels=opt.out_channels
+        )
+        self.discriminator_shadow_to_free = models.Discriminator(
+            input_nc=opt.in_channels
+        )
+        self.discriminator_free_to_shadow = models.Discriminator(
+            input_nc=opt.out_channels
+        )
+
+        # sending models to gpu
+        if torch.cuda.is_available():
+            self.generator_free_to_shadow.cuda()
+            self.generator_shadow_to_free.cuda()
+            self.discriminator_free_to_shadow.cuda()
+            self.discriminator_shadow_to_free.cuda()
+
+        # applying weights init
+        self.generator_free_to_shadow.apply(weights_init)
+        self.generator_shadow_to_free.apply(weights_init)
+        self.discriminator_free_to_shadow.apply(weights_init)
+        self.discriminator_shadow_to_free.apply(weights_init)
+
+        # this criterion uses mean squared error between inputs
+        self.gan_loss_criterion = nn.MSELoss()
+
+        # L1Loss calculates the mean absulute error
+        self.cycle_loss_criterion = nn.L1Loss()
+        self.identity_loss_criterion = nn.L1Loss()
+
+        # optimizer init
+        self.optimizer_gen = self.generator_optimizer(
+            self.generator_shadow_to_free, self.generator_free_to_shadow
+        )
+        self.optimizer_disc_deshadower = self.discriminator_optimizer(
+            self.discriminator_shadow_to_free
+        )
+        self.optimizer_disc_shadower = self.discriminator_optimizer(
+            self.discriminator_free_to_shadow
+        )
+
+        # lr schedulers
+        self.lr_scheduler_gen = torch.optim.lr_scheduler.LambdaLR(
+            self.optimizer_gen,
+            lr_lambda=LR_lambda(opt.n_epochs, opt.epoch, opt.decay_epoch).step,
+        )
+        self.lr_scheduler_disc_s = torch.optim.lr_scheduler.LambdaLR(
+            self.optimizer_disc_shadower,
+            lr_lambda=LR_lambda(opt.n_epochs, opt.epoch, opt.decay_epoch).step,
+        )
+        self.lr_scheduler_disc_d = torch.optim.lr_scheduler.LambdaLR(
+            self.optimizer_disc_deshadower,
+            lr_lambda=LR_lambda(opt.n_epochs, opt.epoch, opt.decay_epoch).step,
```

Comment:
i would consider moving that code to seperate (private) methods

Reply:
We can try that